lib/Service/AvatarService.php

```diff
@@ -0,0 +1,208 @@
+use OC\Files\Filesystem;
+use OCA\Talk\Room;
+use OCP\Files\IAppData;
+use OCP\Files\NotFoundException;
+use OCP\Files\SimpleFS\InMemoryFile;
+use OCP\Files\SimpleFS\ISimpleFile;
+use OCP\Files\SimpleFS\ISimpleFolder;
+use OCP\IAvatarManager;
+use OCP\IConfig;
+use OCP\IL10N;
+use OCP\IURLGenerator;
+use OCP\IUser;
+use OCP\Security\ISecureRandom;
+
+class AvatarService {
+	private IAppData $appData;
+	private IL10N $l;
+	private IConfig $config;
+	private IURLGenerator $url;
+	private ISecureRandom $random;
+	private RoomService $roomService;
+	private IAvatarManager $avatarManager;
+
+	public function __construct(
+		IAppData $appData,
+		IL10N $l,
+		IConfig $config,
+		IURLGenerator $url,
+		ISecureRandom $random,
+		RoomService $roomService,
+		IAvatarManager $avatarManager
+	) {
+		$this->appData = $appData;
+		$this->l = $l;
+		$this->config = $config;
+		$this->url = $url;
+		$this->random = $random;
+		$this->roomService = $roomService;
+		$this->avatarManager = $avatarManager;
+	}
+
+	public function setAvatarFromRequest(Room $room, ?array $file): void {
+		if ($room->getType() === Room::TYPE_ONE_TO_ONE) {
+			throw new InvalidArgumentException($this->l->t('One to one rooms always need to show the other users avatar'));
+		}
+
+		if (is_null($file) || !is_array($file)) {
+			throw new InvalidArgumentException($this->l->t('No image file provided'));
+		}
+
+		if (
+			$file['error'] !== 0 ||
+			!is_uploaded_file($file['tmp_name']) ||
+			Filesystem::isFileBlacklisted($file['tmp_name'])
+		) {
+			throw new InvalidArgumentException($this->l->t('Invalid file provided'));
+		}
+		if ($file['size'] > 20 * 1024 * 1024) {
+			throw new InvalidArgumentException($this->l->t('File is too big'));
+		}
+
+		$content = file_get_contents($file['tmp_name']);
+		unlink($file['tmp_name']);
+		$image = new \OC_Image();
+		$image->loadFromData($content);
+		$image->readExif($content);
+		$this->setAvatar($room, $image);
+	}
+
+	public function setAvatar(Room $room, \OC_Image $image): void {
+		if ($room->getType() === Room::TYPE_ONE_TO_ONE) {
+			throw new InvalidArgumentException($this->l->t('One to one rooms always need to show the other users avatar'));
+		}
+		$image->fixOrientation();
+		if (!($image->height() === $image->width())) {
+			throw new InvalidArgumentException($this->l->t('Avatar image is not square'));
+		}
+
+		if (!$image->valid()) {
+			throw new InvalidArgumentException($this->l->t('Invalid image'));
+		}
+
+		$mimeType = $image->mimeType();
+		$allowedMimeTypes = [
+			'image/jpeg',
+			'image/png',
+		];
+		if (!in_array($mimeType, $allowedMimeTypes)) {
+			throw new InvalidArgumentException($this->l->t('Unknown filetype'));
+		}
+
+
+		$token = $room->getToken();
+		$avatarFolder = $this->getAvatarFolder($token);
+
+		// Delete previous avatars
+		foreach ($avatarFolder->getDirectoryListing() as $file) {
+			$file->delete();
+		}
+
+		$avatarName = $this->random->generate(10, ISecureRandom::CHAR_HUMAN_READABLE);
```

Comment:
Should add the file extension?

```suggestion
		$avatarName = $this->random->generate(6, ISecureRandom::CHAR_HUMAN_READABLE) . ($mimeType === 'image/jpeg' ? '.jpg' : '.png');
```

Reply:
Is possible, but we will need change the current implementation a bit to consider the extension.
For now, the name that we stored on database is only to have a token to expire the previous image. Now, if we send a request without the version, will get the room image too.